README.md
# OnApp CDN Wowza Token Tool

Java and .NET command-line tools for generating tokens for signed streams.

*Note: This tool is mirrored on Github and Bitbucket*


java/changelog.txt
2014-07-24 - version 1.2
==========================================
- Support Java6

2014-07-22 - version 1.1
==========================================
- Stricter constraints for security parameters(expire, ref_allow, ref_deny) values.

2014-07-17 - version 1.0
==========================================
- Fix return result for encryption to 'token=<token_value>'. Append this to the playback URL.
-- Example: http://example.com/video/stream.mp4?token=asdasfoaofo131241o1joosppm

java/src/test/java/com/onapp/cdn/TokenAuthGeneratorTest.java
package com.onapp.cdn;

import static com.onapp.cdn.TokenAuthGenerator.PARAM_EXPIRE;
import static com.onapp.cdn.TokenAuthGenerator.PARAM_REF_ALLOW;
import static com.onapp.cdn.TokenAuthGenerator.PARAM_REF_DENY;
import static com.onapp.cdn.TokenAuthGenerator._encrypt;
import static com.onapp.cdn.TokenAuthGenerator.decrypt;
import static com.onapp.cdn.TokenAuthGenerator.encrypt;
import static com.onapp.cdn.TokenAuthGenerator.main;
import static com.onapp.cdn.TokenAuthGenerator.parseSecurityParameters;
import static com.onapp.cdn.TokenAuthGenerator.validateReferrer;
import static java.lang.String.format;
import static java.util.Arrays.asList;
import static junit.framework.Assert.assertEquals;
import static junit.framework.Assert.assertFalse;
import static junit.framework.Assert.assertTrue;

import java.util.Date;
import java.util.List;
import java.util.Map;

import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.ExpectedException;

/**
 * Unit tests for {@link TokenAuthGenerator}.
 * 
 * @author kwshoo
 * @version 20140722
 */
public class TokenAuthGeneratorTest {
    
    @Rule
    public ExpectedException thrown = ExpectedException.none();
    
    // ======================================================
    // encrypt
    // ======================================================
    @Test
    public void testEncrypt() {
        assertEquals(
                "110ea31ac69c09a2cb7854126719f5d3c3267d24c723eea5cbd99cc4d05426ab679a57015d4e48438c97b921652daec62de3829f8ff437e27449cfdfc2f1e5d9fc47f14e91a51ea7",
                _encrypt("samplekey", "expire=1356955399&ref_allow=*.TrustedDomain.com&ref_deny=Denied.com", false));
    }
    
    @Test
    public void testEncryptUnsupportedParameters() {
        thrown.expect(UnsupportedOperationException.class);
        thrown.expectMessage("Unsupported parameter 'some_unsupported_param'");
        encrypt("samplekey", "some_unsupported_param=123&ref_allow=*.TrustedDomain.com&ref_deny=Denied.com");
    }
    
    @Test
    public void testEncryptEmptyParamValue() {
        thrown.expect(IllegalArgumentException.class);
        thrown.expectMessage("Malformed key/value pair");
        encrypt("samplekey", format("expire=%s&ref_allow=&ref_deny=Denied.com", getFutureDateInSeconds()));
    }
    
    @Test
    public void testEncryptEmptyParamKey() {
        thrown.expect(IllegalArgumentException.class);
        thrown.expectMessage("Malformed key/value pair");
        encrypt("samplekey", "=123&ref_allow=*.TrustedDomain.com&ref_deny=Denied.com");
    }
    
    @Test
    public void testEncryptEmptyParameters() {
        thrown.expect(IllegalArgumentException.class);
        thrown.expectMessage("Parameters must not be empty");
        encrypt("samplekey", "");
    }
    
    // ======================================================
    // parseSecurityParameters
    // ======================================================
    
    @Test
    public void testParseSecurityParameters() {
        long fd = getFutureDateInSeconds();
        Map<String, Object> rt = parseSecurityParameters(format("expire=%s&ref_allow=abc.com,*.TrustedDomain.com,MISSING&ref_deny=Denied.com", fd));
        
        assertEquals(3, rt.size());
        assertEquals(fd, ((Date) rt.get(PARAM_EXPIRE)).getTime() / 1000);
        List<String> refAllowList = asList((String[]) rt.get(PARAM_REF_ALLOW));
        assertEquals(2, refAllowList.size());
        assertTrue(refAllowList.contains("abc.com"));
        assertTrue(refAllowList.contains("*.TrustedDomain.com"));
        List<String> refDenyList = asList((String[]) rt.get(PARAM_REF_DENY));
        assertEquals(1, refDenyList.size());
        assertTrue(refDenyList.contains("Denied.com"));
    }
    
    @Test
    public void testParseSecurityParametersEndWithSymbolAnd() {
        long fd = getFutureDateInSeconds();
        Map<String, Object> rt = parseSecurityParameters(format("&expire=%s&ref_allow=abc.com,*.TrustedDomain.com,&&", fd));
        
        assertEquals(2, rt.size());
        assertEquals(fd, ((Date) rt.get(PARAM_EXPIRE)).getTime() / 1000);
        List<String> refAllowList = asList((String[]) rt.get(PARAM_REF_ALLOW));
        assertEquals(2, refAllowList.size());
        assertTrue(refAllowList.contains("abc.com"));
        assertTrue(refAllowList.contains("*.TrustedDomain.com"));
        assertFalse(rt.containsKey(PARAM_REF_DENY));
    }
    
    @Test
    public void testParseSecurityParametersPastExpire() {
        thrown.expect(IllegalArgumentException.class);
        thrown.expectMessage("Parameter 'expire' should not be a past date");
        long fd = new Date(new Date().getTime() - 99999).getTime() / 1000;
        parseSecurityParameters(format("&expire=%s&ref_allow=abc.com,*.TrustedDomain.com,&&", fd));
    }
    
    @Test
    public void testParseSecurityParametersSpacesBeforeAndAfterReferrer() {
        thrown.expect(IllegalArgumentException.class);
        thrown.expectMessage("Referrer must not be start/end with space(s)");
        parseSecurityParameters(format("&expire=%s&ref_allow= abc.com ,*.TrustedDomain.com,&&", getFutureDateInSeconds()));
    }
    
    @Test
    public void testParseSecurityParametersWildcardReferrer() {
        long fd = getFutureDateInSeconds();
        Map<String, Object> rt = parseSecurityParameters(format("expire=%s&ref_allow=abc.com,*.TrustedDomain.com/Folder1/,", fd));
        
        assertEquals(2, rt.size());
        assertEquals(fd, ((Date) rt.get(PARAM_EXPIRE)).getTime() / 1000);
        List<String> refAllowList = asList((String[]) rt.get(PARAM_REF_ALLOW));
        assertEquals(2, refAllowList.size());
        assertTrue(refAllowList.contains("abc.com"));
        assertTrue(refAllowList.contains("*.TrustedDomain.com/Folder1/"));
        assertFalse(rt.containsKey(PARAM_REF_DENY));
    }
    
    @Test
    public void testParseSecurityParametersInvalidUseOfWildcard() {
        thrown.expect(IllegalArgumentException.class);
        thrown.expectMessage("Wildcard usage(*.DOMAIN) for referrer must exist only at the beginning of a domain");
        parseSecurityParameters(format("expire=%s&ref_allow=abc.com,TrustedDomain.*.com,&", getFutureDateInSeconds()));
    }
    
    private long getFutureDateInSeconds() {
        return new Date(new Date().getTime() + (3600 * 1000)).getTime() / 1000;
    }
    
    // ======================================================
    // validateReferrer
    // ======================================================
    
    @Test
    public void testValidateReferrer() {
        validateReferrer("*.google.com");
        validateReferrer("abc.google.com");
        validateReferrer("google.com/");
        validateReferrer("google.com/video/funny");
        validateReferrer("google.com/video/funny/");
        validateReferrer("google.com/video/funny.mp4");
    }
    
    @Test
    public void testValidateReferrerInvalid() {
        try { validateReferrer(null); }
        catch (IllegalArgumentException e) { assertEquals("Referrer must not be blank", e.getMessage()); }
        try { validateReferrer(""); }
        catch (IllegalArgumentException e) { assertEquals("Referrer must not be blank", e.getMessage()); }
        try { validateReferrer(" *.google.com"); }
        catch (IllegalArgumentException e) { assertEquals("Referrer must not be start/end with space(s)", e.getMessage()); }
        try { validateReferrer("google.com "); }
        catch (IllegalArgumentException e) { assertEquals("Referrer must not be start/end with space(s)", e.getMessage()); }
        try { validateReferrer("google.*.com"); }
        catch (IllegalArgumentException e) { assertEquals("Wildcard usage(*.DOMAIN) for referrer must exist only at the beginning of a domain", e.getMessage()); }
        try { validateReferrer("*.google.*.com"); }
        catch (IllegalArgumentException e) { assertEquals("Wildcard usage(*.DOMAIN) for referrer must exist only at the beginning of a domain", e.getMessage()); }
        try { validateReferrer("http://google.com/video"); }
        catch (IllegalArgumentException e) { assertEquals("Referrer 'http://google.com/video' must not contain protocol", e.getMessage()); }
        try { validateReferrer("google.com/@($2abc/video"); }
        catch (IllegalArgumentException e) { assertEquals("Referrer 'google.com/@*(@$2abc/video' is malformed (RFC 2396)", e.getMessage()); }
        try { validateReferrer("?google.com"); }
        catch (IllegalArgumentException e) { assertEquals("Referrer '?google.com' hostname is invalid(RFC 3490)", e.getMessage()); }
        try { validateReferrer("google>.com/video"); }
        catch (IllegalArgumentException e) { assertEquals("Referrer 'google>.com/video' hostname is invalid(RFC 3490)", e.getMessage()); }
        try { validateReferrer("google.com:3000/video"); }
        catch (IllegalArgumentException e) { assertEquals("Referrer must not contain port number", e.getMessage()); }
    }
    
    // ======================================================
    // decrypt
    // ======================================================
    
    @Test
    public void testDecrypt() {
        assertEquals(
                "expire=1356955399&ref_allow=*.TrustedDomain.com&ref_deny=Denied.com",
                decrypt("samplekey",
                        "110ea31ac69c09a2cb7854126719f5d3c3267d24c723eea5cbd99cc4d05426ab679a57015d4e48438c97b921652daec62de3829f8ff437e27449cfdfc2f1e5d9fc47f14e91a51ea7"));
    }
    
    // ======================================================
    // main
    // ======================================================
    
    @Test
    public void testMainEncrypt() throws Exception {
        main(new String[] { "encrypt", "samplekey", format("expire=%s&ref_allow=*.TrustedDomain.com&ref_deny=Denied.com", getFutureDateInSeconds()) });
    }
    
    @Test
    public void testMainDecrypt() throws Exception {
        main(new String[] { "decrypt", "samplekey", "110ea31ac69c09a2cb7854126719f5d3c3267d24c723eea5cbd99cc4d05426ab679a57015d4e48438c97b921652daec62de3829f8ff437e27449cfdfc2f1e5d9fc47f14e91a51ea7" });
    }
    
    @Test
    public void testMainUnmatchedNumberOfArguments() throws Exception {
        thrown.expect(IllegalArgumentException.class);
        thrown.expectMessage("Expected 3 arguments. Refer to README for usage");
        main(new String[] { "encrypt" });
    }
    
    @Test
    public void testMainInvalidCommand() throws Exception {
        thrown.expect(IllegalArgumentException.class);
        thrown.expectMessage("Invalid action. Refer to README for usage");
        main(new String[] { "invalid", "samplekey", "sample_params" });
    }
    
}

java/src/main/java/com/onapp/cdn/TokenAuthGenerator.java
package com.onapp.cdn;

import static com.google.common.net.InternetDomainName.isValid;
import static java.lang.Long.parseLong;
import static java.lang.String.format;
import static java.util.TimeZone.getTimeZone;
import static org.apache.shiro.codec.Hex.decode;
import static org.apache.shiro.util.StringUtils.hasText;

import java.net.MalformedURLException;
import java.net.URL;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Date;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.StringTokenizer;

import org.apache.shiro.crypto.BlowfishCipherService;
import org.apache.shiro.crypto.CipherService;
import org.apache.shiro.crypto.CryptoException;
import org.apache.shiro.crypto.OperationMode;
import org.apache.shiro.crypto.PaddingScheme;

/**
 * Tool to generate / decrypt token.
 * 
 * @author kwshoo
 * @version 20140723
 */
public class TokenAuthGenerator {
    
    private static CipherService cipherService = new BlowfishCipherService() {{
        setMode(OperationMode.ECB);
        setPaddingScheme(PaddingScheme.PKCS5);
    }};
    
    public static final String PARAM_EXPIRE    = "expire";
    public static final String PARAM_REF_ALLOW = "ref_allow";
    public static final String PARAM_REF_DENY  = "ref_deny";
    
    @SuppressWarnings("serial")
    private static final Set<String> SUPPORTED_PARAMS = new HashSet<String>() {{
        add(PARAM_EXPIRE);
        add(PARAM_REF_ALLOW);
        add(PARAM_REF_DENY);
    }};
    
    public static String encrypt(String key, String parameters) throws CryptoException {
        return _encrypt(key, parameters, true);
    }
    
    /**
     * NOTE: DO NOT USE THIS - FOR TESTING PURPOSE.
     */
    public static String _encrypt(String key, String parameters, boolean isValidateSecurityParams) throws CryptoException {
        parseSecurityParameters(parameters, isValidateSecurityParams);
        return cipherService.encrypt(parameters.getBytes(), key.getBytes()).toHex();
    }
    
    public static String decrypt(String key, String encryptedStr) throws CryptoException {
        return new String(cipherService.decrypt(decode(encryptedStr), key.getBytes()).getBytes());
    }
    
    public static Map<String, Object> parseSecurityParameters(String parameters) {
        return parseSecurityParameters(parameters, true);
    }
    
    /**
     * Given a string of security parameters, it will be parsed and return a map of result.
     * 
     * Example of returned result:
     * <pre>
     *   
     * </pre>
     */
    public static Map<String, Object> parseSecurityParameters(String parameters, boolean isValidateSecurityParams) {
        if (!hasText(parameters))
            throw new IllegalArgumentException("Parameters must not be empty");
        
        StringTokenizer token = new StringTokenizer(parameters, "&");
        Map<String, Object> map = new HashMap<String, Object>();
        while (token.hasMoreTokens()) {
            String[] strArray = token.nextToken().split("=");
            if (strArray.length != 2 || (!hasText(strArray[0]) || !hasText(strArray[1])))
                throw new IllegalArgumentException("Malformed key/value pair");
            
            String paramKey = strArray[0];
            String paramValue = strArray[1];
            
            // Supported param keys check
            if (!isParameterSupported(paramKey))
                throw new UnsupportedOperationException(format("Unsupported parameter '%s'", paramKey));
            
            // Duplication check
            if (!map.containsKey(paramKey)) {
                map.put(paramKey, parse(paramKey, paramValue, isValidateSecurityParams));
            } else {
                throw new IllegalArgumentException(format("Duplicate key '%s' is not allowed", strArray[0]));
            }
        }
        
        return map;
    }
    
    public static Object parse(String key, String value, boolean isValidate) {
        if (PARAM_EXPIRE.equals(key)) {
            long expire = parseLong(value);
            if (isValidate) {
                Calendar _expire = Calendar.getInstance(getTimeZone("UTC"));
                _expire.setTimeInMillis(expire * 1000);
                Calendar _now = Calendar.getInstance(getTimeZone("UTC"));
                if (_now.after(_expire))
                    throw new IllegalArgumentException("Parameter 'expire' should not be a past date");
            }
            return new Date(expire * 1000);
        } else if (PARAM_REF_ALLOW.equals(key) || PARAM_REF_DENY.equals(key)) {
            String refs = value;
            String[] refArr = refs.split(",");
            List<String> refList = new ArrayList<String>();
            for (String ref : refArr) {
                if (ref.equals("MISSING"))
                    continue;
                
                if (isValidate)
                    validateReferrer(ref);
                refList.add(ref);
            }
            return refList.toArray(new String[0]);
        } else {
            throw new UnsupportedOperationException(format("Unsupported parameter '%s'", key));
        }
    }
    
    public static void validateReferrer(String ref) {
        URL url = null;
        String temp = "http://" + ref;
        
        if (!hasText(ref)) {
            throw new IllegalArgumentException("Referrer must not be blank");
        }
        
        if (ref.startsWith(" ") || ref.endsWith(" "))
            throw new IllegalArgumentException("Referrer must not be start/end with space(s)");
        
        if (ref.contains("*")) {
            if (!ref.startsWith("*.") || ref.lastIndexOf("*") > 0) {
                throw new IllegalArgumentException("Wildcard usage(*.DOMAIN) for referrer must exist only at the beginning of a domain");
            }
            
            // So we can test the ref validity with URL constructor =P
            temp = temp.replace("*", "temp");
        }
        
        // Fail if referrer contains protocol
        try {
            url = new URL(ref);
            if (hasText(url.getProtocol())) {
                throw new IllegalArgumentException(format("Referrer '%s' must not contain protocol", ref));
            }
        } catch (MalformedURLException e) {}
        
        try {
            url = new URL(temp);
        } catch (MalformedURLException e) {
            throw new IllegalArgumentException(format("Referrer '%s' is malformed (RFC 2396)", ref));
        }
        
        if (!isValid(url.getHost())) {
            throw new IllegalArgumentException(format("Referrer '%s' hostname is invalid(RFC 3490)", ref));
        }
        
        if (url.getPort() != -1) {
            throw new IllegalArgumentException(format("Referrer must not contain port number", ref));
        }
    }
    
    public static boolean isParameterSupported(String parameterKey) {
        return SUPPORTED_PARAMS.contains(parameterKey); 
    }
    
    public static void main(String[] args) throws Exception {
        if (args.length != 3) 
            throw new IllegalArgumentException("Expected 3 arguments. Refer to README for usage");
        if (!"encrypt".equals(args[0]) && !"decrypt".equals(args[0])) 
            throw new IllegalArgumentException("Invalid action. Refer to README for usage");
        
        if ("encrypt".equals(args[0])) {
            String key = args[1];
            String parameters = args[2];
            String encryptedStr = encrypt(key, parameters);
            System.out.println("token=" + encryptedStr);
            return;
        } else if ("decrypt".equals(args[0])) {
            String key = args[1];
            String encryptedStr = args[2];
            System.out.println("security parameters=" + decrypt(key, encryptedStr));
            return;
        }
    }
}
